### ISSUE TITLE
bun dedupe needed

### ISSUE THREAD

**Issue by imattaai** [oven-sh/bun#8594]
### What is the problem this feature would solve?

When working with expo and react-native (or other JS projects)
Package mismatch can lead to unexplained errors in the console.
Several threads recommend using 'npm dedupe' which I haven't found an alternative for.
What's worked for me is setting package "overrides" in package.json from recommendations and guessing the package.

### What is the feature you are proposing to solve the problem?

Unlinke yarn.lock where you can see packages mentioned and versions, I would like a way to know the different versions of packages in my app so I can set an override or take alternative measures (upgrade/downgrade packages).

### What alternatives have you considered?

Alternatives now are reverting to npm or guessing the overriding package from my current knowledge.

**Comment by ibash** [comment]
If you run `bun bun.lockb` you'll get a print out of what packages and what dependencies are being resolved to. It's a bit manual but it could help you track down conflicting packages.

**Comment by ArnaudBarre** [comment]
Same as #1343

**Comment by Electroid** [comment]
Duplicate of #1343

**Comment by Electroid** [comment]
This is something we plan on implementing.

**Comment by NikhilVerma** [comment]
This is somewhat related to nested overrides/resolutions. As most of the dedupe issues can be fixed by a nested resolution enforcement. But right now if I just force bun to use a specific version of a dependency. It still installs other meta-dependencies of the same name but with different versions

**Comment by Jarred-Sumner** [comment]
Implemented! Copy-pasting comment from https://github.com/oven-sh/bun/issues/1343

This shipped in #38333.

`bun dedupe` reads `bun.lock`, collapses every package locked at more than one version onto a version that satisfies all of its ranges (newest preferred), and installs. Your esbuild example (`^0.15.7` + `^0.15.8`) collapses onto one version.

```sh
$ bun dedupe
- esbuild@0.15.10
- react@18.2.0
Removed 2 duplicate versions

$ bun dedupe --check   # exits 1 if anything is removable, for CI
🎉 No duplicates — checked 148 packages, every one already resolves to a single version [9.00ms]
```

- No registry requests, no widened ranges, `package.json` untouched; overrides, catalogs, patches, workspaces, and both linkers are handled (nested `node_modules` copies are deleted, no `rm -rf`).
- Only versions already in the lockfile are used; `bun update <name>` now re-resolves that package everywhere, so routine bumps end up on one copy.

It'll be in the next canary (`bun upgrade --canary`) and in Bun v1.4 — if it doesn't cover your case, comment here.

*(This comment was written by Claude, on behalf of the Bun team.)*
